### PR TITLE
Add Exception Middleware

### PR DESCRIPTION
**Pull request by crobibero**
Adds Exception Middleware for graceful handling of exceptions from Jellyfin.Api.

Part of #2872

### CODE REVIEW

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            {
+                return string.Empty;
+            }
+
+            // Strip any information we don't want to reveal
+            msg = msg.Replace(
+                _configuration.ApplicationPaths.ProgramSystemPath,
+                string.Empty,
+                StringComparison.OrdinalIgnoreCase);
+            msg = msg.Replace(
```
Comment (cvium) [inline Jellyfin.Server/Middleware/ExceptionMiddleware.cs:112]: You can chain `.Replace` calls.

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                if (context.Response.HasStarted)
+                {
+                    _logger.LogWarning("The response has already started, the exception middleware will not be executed.");
+                    throw;
+                }
+
+                ex = GetActualException(ex);
+                _logger.LogError(ex, "Error processing request: {0}", ex.Message);
+                context.Response.StatusCode = GetStatusCode(ex);
+                context.Response.ContentType = "text/plain";
```
Comment (cvium) [inline Jellyfin.Server/Middleware/ExceptionMiddleware.cs:59]: https://docs.microsoft.com/en-us/dotnet/api/system.net.mime.mediatypenames.text.plain?view=netcore-3.1

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            catch (Exception ex)
+            {
+                if (context.Response.HasStarted)
+                {
+                    _logger.LogWarning("The response has already started, the exception middleware will not be executed.");
+                    throw;
+                }
+
+                ex = GetActualException(ex);
+                _logger.LogError(ex, "Error processing request: {0}", ex.Message);
```
Comment (cvium) [inline Jellyfin.Server/Middleware/ExceptionMiddleware.cs:57]:
```suggestion
                _logger.LogError(ex, "Error processing request: {ExceptionMessage}", ex.Message);
```

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExceptionMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">Next request delegate.</param>
+        /// <param name="loggerFactory">Instance of the <see cref="ILoggerFactory"/> interface.</param>
+        /// <param name="serverConfigurationManager">Instance of the <see cref="IServerConfigurationManager"/> interface.</param>
+        public ExceptionMiddleware(
+            RequestDelegate next,
+            ILoggerFactory loggerFactory,
```
Comment (cvium) [inline Jellyfin.Server/Middleware/ExceptionMiddleware.cs:29]: Shouldn't this be `ILogger<ExceptionMiddleware>`?

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    throw;
+                }
+
+                ex = GetActualException(ex);
+                _logger.LogError(
+                    ex,
+                    "Error processing request: {ExceptionMessage}. URL {Method} {Url}. ",
+                    ex.Message,
+                    context.Request.Method,
+                    context.Request.Path);
```
[the remark this answers is not in the document]
Comment (crobibero) [reply Jellyfin.Server/Middleware/ExceptionMiddleware.cs]: The implementation I have on disk is somehow different, updated to this version.

File: Jellyfin.Server/Middleware/ExceptionMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                        ex.Message.TrimEnd('.'),
+                        context.Request.Method,
+                        context.Request.Path);
+                }
+                else
+                {
+                    _logger.LogError(
+                        ex,
+                        "Error processing request. URL {Method} {Url}.",
+                        ex.Message.TrimEnd('.'),
```
Comment (mark-monteiro) [inline Jellyfin.Server/Middleware/ExceptionMiddleware.cs:82]: Don't think you want this ;)

Comment (crobibero) [reply Jellyfin.Server/Middleware/ExceptionMiddleware.cs]: image

**Pull request merged**